### DISCUSSION TITLE
New Release

### DISCUSSION THREAD

**Discussion by alejandrogallo** [papis/papis#799] [General]
Hi all,

@alexfikl, @jghauser 
how is it looking to prepare a new release for version 0.14?
It think many of the features we wanted to include are slowly coming to form, is it not?

**Comment by jghauser** [comment]
Yeah, absolutely! I think we should do a release. Someone in #311 offered to make the `clean_document_name()` function configurable and expressed wanting to get this into the next release, though it's been a couple of weeks now. I'll ping them in that thread now. Other than that, I think it would be a good moment to cut a new release :grinning:

**Comment by alexfikl** [comment]
I agree! We should definitely make a release!

I was thinking of also getting #745 in so that the new Windows effort from #780 works better. Is that ok? It's tested a bit on other platforms, so it should be fine.

**Comment by alexfikl** [comment]
I was just looking through some Python 3.13 release notes and remembered they're removing the `cgi` module from the standard library. We're using only `cgi.FieldStorage` in `papis/commands/serve.py`.

Python 3.13 will be out in October-ish, but it would be nice to be compatible with it in this release too.

EDIT: Added it to the CI to see if anything else breaks #807.

**Reply by alejandrogallo** [reply]
Thank you for taking a look and noticing.

I was thinking about just depending on Flask, since
right now the serve command is quite limited in the sense that it allows only one thread, so it would be nice to be able to serve several requests at the same time, and AFAIK Flask is very minimal in terms of dependency load. What do you think? I could rewrite quickly `serve.py` in Flask in order to support this, since I use a lot the webapp in general. I also wanted to implement a simple user login for the webapp.

**Reply by alexfikl** [reply]
Yeah, I'm definitely fine with that. Flask is pretty small and generally available, so shouldn't be a big deal!

Depending on how fast we want to push out a release now, we can always make another small release in the fall to fix that though.

**Reply by alexfikl** [reply]
Also, just throwing it out there, maybe for the next release:

The `serve` command is growing quite big.. Is it worth spinning it out to an external plugin?
I'm not completely for the idea because external plugin always means more pain when installing and the webapp seems very useful for anyone who isn't reasonably comfortable with the commandline.

**Reply by alejandrogallo** [reply]
you're right, I would say that the webapp will grow, but thankfully we have no serious amount of css or js, but it might be wanted in the future. 
In the case that we spin it out as an external plugin, say papis-web,
we can always add it as a compulsory dependency and the user will not notice the difference, right ? It would be a bit painful from the side of the developers, but not many people are contributing there anyways AFAIK

**Reply by alexfikl** [reply]
Yeah, good point! That can probably be nicely solved with some smarter packaging/dependencies.

**Comment by alexfikl** [comment]
Both #745 and #803 (fixing #311) are in now! :tada:

**Comment by kiike** [comment]
What features/bugs would you like to work on before the release? I'd like to offer my help.

For instance, will #711 get in? If not, I can myself to port the default rules into jinja2 for those so that those who want to try jinja2 can use the picker without having to redefine all values.

**Reply by alexfikl** [reply]
I don't think we have anything in particular on the roadmap. It's mostly up to when @alejandrogallo has some time do make sure everything still works and do the release :grin: 

> For instance, will https://github.com/papis/papis/pull/711 get in? 

As far as #711 is concerned, it seems like several people have tested it out and all the reported issues should be fixed, so it's in a pretty good state to be merged. @alejandrogallo What do you think?

> If not, I can myself to port the default rules into jinja2 for those so that those who want to try jinja2 can use the picker without having to redefine all values.

That should be quite useful regardless! Maybe we can add it as an example in the docs if nothing else. It would also help to see what the jinja formats look like, so that we can maybe add some additional filters to help out with common usecases.

**Reply by kiike** [reply]
> As far as #711 is concerned, it seems like several people have tested it out and all the reported issues should be fixed, so it's in a pretty good state to be merged. @alejandrogallo What do you think?

OK! I'll translate the default rules after the PR I am preparing.

> That should be quite useful regardless! Maybe we can add it as an example in the docs if nothing else. It would also help to see what the jinja formats look like, so that we can maybe add some additional filters to help out with common usecases.

That's true! It will be good to see the possibilities even if just for showcase sake!

**Comment by jghauser** [comment]
Should we do a release finally? :) Also, what do you think about using [release-please](https://github.com/googleapis/release-please) so that we can release a bit more often and more easily? I've recently started using it for papis.nvim and while it has some fiddliness, it is quite convenient!

**Reply by alexfikl** [reply]
Yeeees! :tada: 

I've just been using https://github.com/pypa/gh-action-pypi-publish to push releases to PyPI every time a new tag is created in the repository. That seems to work pretty well, but it doesn't handle the changelog like `release-please` seems to do? In general we should be able to streamline the release quite a bit, since we're just a nice pure Python package :grin:

**Reply by jghauser** [reply]
yeah what i do for papis.nvim is that release-please handles publishing releases on github and then there's another action that creates a luarock on luarock.org whenever there's a new release. i'm sure something like this would work for python packages also.

**Reply by alexfikl** [reply]
That's very nifty! I'll probably try and play with improving the release process for the next release (if no one else beats me to it) :grin: We've gathered quite a lot of changes both for 0.13 and 0.14, so it would definitely be nice to make more frequent releases.

**Reply by ExpandingMan** [reply]
Just want to chime in to say that a new release would be very much appreciated.  Python package management makes it a lot harder to use this package than it would otherwise be, and not having a recent release compounds that issue.

**Comment by ExpandingMan** [comment]
Bump. :slightly_smiling_face:

**Comment by alexfikl** [comment]
@alejandrogallo Any chance we can make a release? :grin: Things seem to have calmed down with any big features lately, so it seems like a good time.

**Comment by tuurep** [comment]
Bump

What's in the way? If I can do anything to help, I'm interested :)

Very much agreed with @ExpandingMan, the packaging situation is frictionful.

**Reply by alexfikl** [reply]
I'm afraid it's just "real life stuff". We're aware we've dragged this out a bit too much at this point and are trying to get a more automatic release deployment thing going :(

**Reply by tuurep** [reply]
Thanks, yes the release improvement plans from above are great :)

At the moment I'm coping by using [`papis-git`](https://aur.archlinux.org/packages/papis-git) from the AUR but it constantly requires some troubleshooting/manual intervention

Wish I could use `pipx` to install papis, but target the latest commit, if you know what I mean :smile:

**Reply by alexfikl** [reply]
> Wish I could use pipx to install papis, but target the latest commit, if you know what I mean 😄

Does something like the examples from https://pipx.pypa.io/stable/examples/ not work? It seems like doing this should do the trick
```
pipx install git+https://github.com/papis/papis@main
```
(I have zero experience with pipx and also use the `papis-git` PKGBUILD with slight modifications)

**Reply by tuurep** [reply]
Ah, I'm just now learning about this

Looks promising, though I have some problems with what looks to be mostly with trying to use jinja2 as formatter, it not using pacman's `python-jinja` and pipx warning me that pipx shouldn't be used to install libraries if I try to use that.

Could the `pipx inject` (https://github.com/papis/papis/discussions/867#discussioncomment-9689683) work in this context as well?

Full error log running `papis open` if you're interested in it:

<details>

```
$ papis open
Error while loading entrypoint 'EntryPoint(name='isbn', value='papis.isbn:Importer', group='papis.importer')': No module named 'pkg_resources'.
Error while loading entrypoint 'EntryPoint(name='isbn', value='papis.isbn:explorer', group='papis.explorer')': No module named 'pkg_resources'.
[INFO] database.cache: Indexing library. This might take a while...
[ERROR] format: The 'jinja2' formatter requires the 'jinja' library. To use this functionality install it using e.g. 'pip install jinja2'. (Caught exception 'ModuleNotFoundError: No module named 'jinja2''. Use `--log DEBUG` to see traceback)
[ERROR] format: Invalid formatter 'jinja2'. Registered formatters are 'jinja2', 'python'. (Caught exception 'ModuleNotFoundError: No module named 'jinja2''. Use `--log DEBUG` to see traceback)
Traceback (most recent call last):
  File "/home/tuure/.local/pipx/venvs/papis/lib/python3.12/site-packages/papis/format.py", line 270, in get_formatter
    FORMATTER = mgr[name].plugin()
                ^^^^^^^^^^^^^^^^^^
  File "/home/tuure/.local/pipx/venvs/papis/lib/python3.12/site-packages/papis/format.py", line 215, in __init__
    raise exc
  File "/home/tuure/.local/pipx/venvs/papis/lib/python3.12/site-packages/papis/format.py", line 209, in __init__
    import jinja2       # noqa: F401
    ^^^^^^^^^^^^^
ModuleNotFoundError: No module named 'jinja2'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/tuure/.local/bin/papis", line 8, in <module>
    sys.exit(run())
             ^^^^^
  File "/home/tuure/.local/pipx/venvs/papis/lib/python3.12/site-packages/click/core.py", line 1157, in __call__
    return self.main(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tuure/.local/pipx/venvs/papis/lib/python3.12/site-packages/click/core.py", line 1078, in main
    rv = self.invoke(ctx)
         ^^^^^^^^^^^^^^^^
  File "/home/tuure/.local/pipx/venvs/papis/lib/python3.12/site-packages/click/core.py", line 1688, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tuure/.local/pipx/venvs/papis/lib/python3.12/site-packages/click/core.py", line 1434, in invoke
    return ctx.invoke(self.callback, **ctx.params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tuure/.local/pipx/venvs/papis/lib/python3.12/site-packages/click/core.py", line 783, in invoke
    return __callback(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tuure/.local/pipx/venvs/papis/lib/python3.12/site-packages/papis/commands/open.py", line 174, in cli
    documents = papis.cli.handle_doc_folder_query_all_sort(query,
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tuure/.local/pipx/venvs/papis/lib/python3.12/site-packages/papis/cli.py", line 156, in handle_doc_folder_query_all_sort
    documents = handle_doc_folder_query_sort(query,
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tuure/.local/pipx/venvs/papis/lib/python3.12/site-packages/papis/cli.py", line 134, in handle_doc_folder_query_sort
    documents = handle_doc_folder_or_query(query, doc_folder)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tuure/.local/pipx/venvs/papis/lib/python3.12/site-packages/papis/cli.py", line 116, in handle_doc_folder_or_query
    return papis.database.get().query(query)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tuure/.local/pipx/venvs/papis/lib/python3.12/site-packages/papis/database/cache.py", line 216, in query
    docs = self.get_documents()
           ^^^^^^^^^^^^^^^^^^^^
  File "/home/tuure/.local/pipx/venvs/papis/lib/python3.12/site-packages/papis/database/cache.py", line 148, in get_documents
    self.documents = papis.utils.folders_to_documents(folders)
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tuure/.local/pipx/venvs/papis/lib/python3.12/site-packages/papis/utils.py", line 333, in folders_to_documents
    result = parmap(papis.document.from_folder, folders)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tuure/.local/pipx/venvs/papis/lib/python3.12/site-packages/papis/utils.py", line 93, in parmap
    _ = papis.format.get_formatter()
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tuure/.local/pipx/venvs/papis/lib/python3.12/site-packages/papis/format.py", line 276, in get_formatter
    raise InvalidFormatterError(f"Invalid formatter: '{name}'")
papis.format.InvalidFormatterError: Invalid formatter: 'jinja2'
```

</details>

**Reply by tuurep** [reply]
Yes! this did work great:

```
$ pipx inject papis setuptools
$ pipx inject papis jinja2
```

Very convenient. Thanks for the help, when I didn't expect to be helped haha :smile:

**Edit:** 

Updating using `pipx upgrade papis` when there's a new commit remains a pain: `papis is already at latest version 0.13`, will see if there's a way to upgrade without having to reinstall.

**Comment by ExpandingMan** [comment]
Perhaps this thread is not particularly visible to maintainers?  Would it be unreasonable to open an issue?

**Reply by alexfikl** [reply]
I'm afraid real life is taking up a bit too much time at the moment, but we're trying to smooth out the release process so hopefully it'll go better in the future.

No need for opening an issue :grin:

**Comment by ExpandingMan** [comment]
I frankly was going to ditch `papis` for alternatives because I find it pretty onerous to get latest main running.  Again I see this as a python package management issue: because python package management is so awful doing simple things is hard.  Even installing releases from `pip` isn't so great the way python in linux works these days.

However, I have found that `papis` seems vastly superior to any alternative I can find.  Suffice it to say, a new release would still be hugely appreciated :sweat_smile:

(Hopefully you don't think I'm being too obnoxious by asking yet again.  I would have kept my mouth shut if I did not see so many commits happening :slightly_smiling_face: )

**Reply by tuurep** [reply]
I can relate :smile: even at some point I tried if I could implement my own minimalistic bib manager, but the more I got into it, I saw that this is what papis already is.

(the fzf integration outofthebox is a huge part of why it works so well for me though)

AUR `papis-git` has a new problem, for an unknown reason `python-sphinx-design (extra)` package no longer exists. There's an AUR package `python-sphinx_design` (note: underscore) however.

I left a comment on [`papis-git`](https://aur.archlinux.org/packages/papis-git).

**Reply by alexfikl** [reply]
We finally managed to get together and push out a new release! Hopefully it'll be less of a hassle in the future. :tada:

**Reply by tuurep** [reply]
This is fantastic :) thanks!
